### PR TITLE
bpo-43468: Per instance locking for  functools.cached_property 

### PR DESCRIPTION
**Pull request by rhettinger**
Draft PR.  It needs a NEWS entry, perhaps more tests, and perhaps some more factoring.

https://bugs.python.org/issue43468

### CODE REVIEW

**Comment by sweeneyde** [comment]
It looks like `threading._register_atexit` could lazily import `functools` to avoid the circular `threading <=> functools` import

File: Lib/functools.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 from collections import namedtuple
 # import types, weakref  # Deferred to single_dispatch()
 from reprlib import recursive_repr
-from _thread import RLock
+from threading import RLock, Condition, get_ident
```
Comment (ambv) [inline Lib/functools.py:21]: This change introduces an import cycle which causes tests to fail.

File: Misc/NEWS.d/next/Library/2021-08-04-18-30-25.bpo-43468.PlU--J.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Fixed locking for functools.cached_property() to have instance specific
+locking.  Formerly, it blocked all instances of the same class, making it
+suitable for parallel I/O bound computations (which were a key use case).
```
Comment (ptmcg) [inline Misc/NEWS.d/next/Library/2021-08-04-18-30-25.bpo-43468.PlU--J.rst]: Do you mean "unsuitable"?

File: Lib/functools.py (hunk 1 above)
Comment (eric-nieuwland) [reply Lib/functools.py:21]: There is a race condition on 980-983.
Scenario:
- 3 threads, 1 registered in `updaters`
- 2nd and 3rd will wait
- *both* waiting threads will see registration of registered thread disappear
Mutual exclusion breaks.

File: Misc/NEWS.d/next/Library/2021-08-04-18-30-25.bpo-43468.PlU--J.rst (hunk 2 above)
Comment (rhettinger) [reply Misc/NEWS.d/next/Library/2021-08-04-18-30-25.bpo-43468.PlU--J.rst]: Yes.  Fixed now.

**Comment by Erotemic** [comment]
Can someone confirm or deny if my solution on twitter here: 
https://twitter.com/erotemic/status/1431081138332835844

with a MWE here: 
https://twitter.com/erotemic/status/1431118636568268800

I also made a PR here: https://github.com/rhettinger/cpython/pull/7

I suppose one test is to see if the dashboards pass.

**Pull request closed without merging**